x/wasm/keeper/keeper.go

```diff
@@ -1213,3 +1218,68 @@ func (h DefaultWasmVMContractResponseHandler) Handle(ctx sdk.Context, contractAd
 	}
 	return result, nil
 }
+
+// GetByteCodeByChecksum queries the wasm code by checksum
+func (k Keeper) GetByteCodeByChecksum(ctx sdk.Context, checksum wasmvm.Checksum) ([]byte, error) {
+	return k.wasmVM.GetCode(checksum)
+}
+
+// iteratePinnedCodes iterates over all pinned code ids in ascending order
+func (k Keeper) iteratePinnedCodes(ctx sdk.Context, cb func(codeID uint64) bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.PinnedCodeIndexPrefix)
+	iter := store.Iterator(nil, nil)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		codeID := sdk.BigEndianToUint64(iter.Key())
+		if cb(codeID) {
+			return
+		}
+	}
+}
+
+// PruneWasmCodes deletes code info for unpinned codes <= than latestCodeID
+func (k Keeper) PruneWasmCodes(ctx sdk.Context, latestCodeID uint64) error {
+	usedCodeIds := make(map[uint64]struct{}, 0)
+	usedChecksums := make(map[string]bool, 0)
+
+	// collect code ids used by contracts
+	k.IterateContractInfo(ctx, func(_ sdk.AccAddress, info types.ContractInfo) bool {
+		usedCodeIds[info.CodeID] = struct{}{}
+		return false
+	})
+
+	// collect pinned code ids
+	k.iteratePinnedCodes(ctx, func(codeID uint64) bool {
+		usedCodeIds[codeID] = struct{}{}
+		return false
+	})
+
+	// check if instances are used only by unpinned code ids <= latestCodeID
+	k.IterateCodeInfos(ctx, func(codeID uint64, info types.CodeInfo) bool {
+		if _, ok := usedCodeIds[codeID]; !ok && codeID <= latestCodeID {
```

Comment:
The `codeID <= latestCodeID` is not correct. Just remove this part that we iterate over the whole set of code infos

For example, there can be code ID 1 and 10 for the same checksum. If only code ID 10 has an instance then a prune with max code id 2 would miss the instance.

Reply:
I iterate over the whole set of code infos. I need this condition to mark as "false" only the checksums with unpinned code ids <= latestCodeID. This helps because we cannot delete checksums with unpinned code ids > latestCodeID.
These will be marked as "true", so as usedChecksums.. same as if they were pinned.
This also answer to the other comment why we need map[string]bool and it is not enough to have map[string]struct{}